---
output:
  BiocStyle::html_document
---

<!---
Because we split the analysis pipeline in different independent files, to speed up processing it, here in the setup block we load libraries and objects that were loaded or produced in the previously processed file, and which are necessary in this file.
--->

```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(geneplotter)
opts_chunk$set(cache=TRUE,
               cache.path="cache/file2",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")
se.unpaired <- readRDS(file.path("rawCounts", "seLUSC.rds"))
dg.unpaired <- readRDS(file.path("results", "dge.rds"))
se <- readRDS(file.path("results", "se.paired.rds"))
dge <- readRDS(file.path("results", "dge.paired.rds"))
se.filt.unnorm <- readRDS(file.path("results", "se.paired.filt.unnorm.rds"))
dge.filt.unnorm <- readRDS(file.path("results", "dge.paired.filt.unnorm.rds"))
se.filt <- readRDS(file.path("results", "se.paired.filt.rds"))
dge.filt <- readRDS(file.path("results", "dge.paired.filt.rds"))
```

# Differential expression

## Differential expression using limma-voom and SVA

A differential expression analysis is performed using the [limma](http://bioconductor.org/packages/release/bioc/html/limma.html) and [sva](http://bioconductor.org/packages/sva.html) R/Bioconductor packages.

To use this method, two model matrices are generated: a matrix of the model including the outcome of interest and adjustment variables, and a matrix of the null model that only includes adjustment variables. The outcome of interest is the type of sample, and as the analysis is paired (both types of samples come from the same patient) a unique identifier of each patient is included. Finally, as there are not known factors to adjust, an intercept term of 1 is given to the null model.

Even if no known sources of variation were found, surrogate variables representing unknown sources of variation can be added to the model. Those can be estimated with the `sva` package, and later be appended to the model.

Finally, `voom` is used to fit the model, and the moderated t-statistics are calculated.

```{r, message=FALSE}
library(sva)
# Obtain the patient ID, as the analysis is paired
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)

# Build the model (mod) and null (mod0) matrices 
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))

# Estimate surrogate variables
sv <- sva(assays(se.filt)$logCPM, mod = mod, mod0 = mod0)

#pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
#sum(p.adjust(pv, method="fdr") < 0.01)

# Add surrogates to the design matrix
len <- length(colnames(mod))
mod <- cbind(mod, sv$sv)
colnames(mod) <- c(colnames(mod)[1:len], paste0("SV", 1:sv$n))
v <- voom(dge.filt, mod, plot=TRUE)

fit <- lmFit(v, mod) # Fit the model with the voom weights
fit <- eBayes(fit) # Calculate moderated t-statistics
```

`sva` was able to estimate `r sv$n.sv` surrogate variables. Regarding the mean-variance plot obtained with `voom`, it shows the variance in the y axis and the mean of expression in the x axis for each gene, represented as a black dot. Such plots contain a red curve that is fitted to the dots and show trends in the data. Normally increases in expression levels mean a decrease in the variance, until a plateau is reached for highly expressed genes. This trens is observed in the plot obtained for this data. Moreover, the plot is also a diagnosis of filtering steps performed upstream. The fact that there is not a drop in variance for the lower expression values indicates that the filtering threshold was good.

<!--Info about voom plot from https://f1000research.com/articles/5-1408, https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29 and https://stats.stackexchange.com/questions/160255/voom-mean-variance-trend-plot-how-to-interpret-the-plot-->

As we are performing multiple tests, the probability of type I errors (rejecting the null hypotheses when it is true). There are multiple statistical tools to decrease this errors, such as the false discovery rate (FDR). This technique consists on defining an acceptable rate of type I errors (false discoveries). Here, a FDR of 0.1 is set, which means that 1% of the rejections of the null hypothesis will be incorrect.

```{r}
FDRcutoff <- 0.1
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

Gene metadata such as the gene symbol and the chromosome can be added for an easier interpretation of the obtained statistics, and to observe the chromosome distribution of the differentially expressed genes.

```{r}
# Get gene info...
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)

# and add it to the results
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 20)
```

This table contains the relevant statistics used in the analysis, such as raw and adjusted p-values. In the case of the later ones, it can be seen that their values are close to one, which means that the result is not significant. The number of differentially expressed genes under the `r FDRcutoff` FDR cuttoff can be checked by applying a filter to this table.

```{r}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
```

The differential expression analysis results in `r length(rownames(tt)[tt$adj.P.Val < FDRcutoff]))` differentially expressed genes.

#### Diagnostics

The distributions of raw p-values and moderated t-statistics can be used as a diagnosis of the tests performed.

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

Under the null hypothesis, which is that most genes are not differentially expressed, the raw p-value distribution should be uniform with a peak at low p-values for the truly DE genes. It is not the case.

Regarding the second plot, it represents the quantiles of our data sample against the theoretical distribution they should follow. The diagonal represents the null hypothesis. Most genes are off-diagonal, indicating that differentially expressed genes exist, but they are still close to it because they are not significant.

## Alternative differential expression analysis approaches

As said previously, adding surrogate variables adjusts for unknown variation, which in practice would translate into an increased number of differentially expressed genes with respect to analyses performed without including them. Usually this unknown variation would account for non-biological variability, but in some cases the surrogate variables can capture biological variability that is important for the outcome. One of such cases happens when there is some sort of heterogeneity in the data, such as unknown subgroups. When this happens, `sva` is not reccomended.

Following this idea, two different modifications to the surrogate variables before performing the diferential expression analysis were done. In the first case they were not included, while only one of them was removed in the second. Note that code that was identical to the first analysis will not be displayed for clarity.

### Differential expression analysis without surrogate variable adjustment

The first analysis will consist on the removal of the surrogate variable estimation and inclusion into the model.

```{r, message=FALSE}
# Obtain the patient ID, as the analysis is paired
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)

# Build the model (mod) and null (mod0) matrices 
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))

v <- voom(dge.filt, mod, plot=TRUE)

fit <- lmFit(v, mod) # Fit the model with the voom weights
fit <- eBayes(fit) # Calculate moderated t-statistics
```

`sva` was able to estimate `r sv$n.sv` surrogate variables. There are not relevant variations in the mean-variance plot.

The correction for multiple testing is the same as in the previous analysis, a FDR of 0.1. After this, gene metadata is added for interpretation purposes.

```{r echo=FALSE}
FDRcutoff <- 0.1
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

```{r echo=FALSE}
# Get gene info...
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)

# and add it to the results
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 20)
```

A drop on the adjusted p-values is already observed, leading to significant expression changes. The number of significantly differentially expressed genes is obtained by filtering the table by the `r FDRcutoff` FDR cuttoff value.

```{r echo=FALSE}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
```

In this case, the differential expression analysis results in `r length(rownames(tt)[tt$adj.P.Val < FDRcutoff]))` significant differentially expressed genes. Now, the chromosome distribution of the genes can be assessed.

```{r}
sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE)
plot(sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE))
axis(side=1, cex.axis=0.35)
```

The distribution is pretty variable: most differentially expressed genes are located in big chromosomes, but smaller chromosomes are sometimes located higher in the ranking from what would be expected by its size.

A list of the top 10 differentially expressed genes can be obtained:

```{r}
fit$genes$symbol[top]
```

#### Diagnostics

The performance of the statistical test is diagnosed with the distributions of raw p-values and moderated t-statistics.

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

The assumption of a flat distribution of raw p-values for the non-DE genes under the null hypothesis holds in this case. In the QQ plot, more genes are far from the null hypothesis, which indicates significant differential expression.

The results are diagnosed with volcano and MA plots, in which the top 10 differentially expressed genes are highlighted.

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 3, 2))
volcanoplot(fit, coef = 2, highlight = 10, names = fit$genes$symbol, main = "Volcano")

top <- order(fit$lods[, 2], decreasing = TRUE)[1:10]
limma::plotMA(fit, coef = 2, status = rownames(fit$lods) %in% DEgenes, legend = FALSE,
main = "MA Plot", hl.pch = 46, hl.cex = 4, bg.pch = 46, bg.cex = 3, las = 1)
text(fit$Amean[top], fit$coef[top, 2], fit$genes$symbol[top], cex = 0.5, pos = 4)
```

### Differential expression analysis removing the first surrogate variable

The removal of all the surrogate variables might be considered radical, as some of them might be capturing part of the unwanted non-biological variation. It is possible that only one of the variables is capturing the relevant biological information. By removing each of them one at a time, it was observed that the variable responsible for this was the first one, so the following analysis is performed by removing it. The others can be done by simply modifying the number on the `svmod <- sv$sv[,-1]` line.

#### Analysis

```{r, message=FALSE}
library(sva)
# Obtain the patient ID, as the analysis is paired
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)

# Build the model (mod) and null (mod0) matrices 
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))

# Estimate surrogate variables
sv <- sva(assays(se.filt)$logCPM, mod = mod, mod0 = mod0)

# Remove the first surrogate variable
svmod <- sv$sv[,-1]

# Add surrogates to the design matrix
len <- length(colnames(mod))
mod <- cbind(mod, svmod)
colnames(mod) <- c(colnames(mod)[1:len], paste0("SV", 1:(sv$n-1)))
v <- voom(dge.filt, mod, plot=TRUE)

fit <- lmFit(v, mod) # Fit the model with the voom weights
fit <- eBayes(fit) # Calculate moderated t-statistics
```

`sva` estimated `r sv$n.sv` surrogate variables, though `r (sv$n.sv)-1` will be used. There are not relevant variations in the mean-variance plot.

The correction for multiple testing is maintained to a FDR of 0.1. Some preliminary data on upregulation and downregulation is obtained and gene metadata is added for interpretation purposes.

```{r echo=FALSE}
FDRcutoff <- 0.1
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

```{r echo=FALSE}
# Get gene info...
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)

# and add it to the results
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 20)
```

As in the study without surrogate variables, a drop on the adjusted p-values is already observed, leading to significant expression changes. The number of significantly differentially expressed genes is obtained by filtering the table by the `r FDRcutoff` FDR cuttoff value.

```{r echo=FALSE}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
```

So the differential expression analysis without the first surrogate variable results in `r length(rownames(tt)[tt$adj.P.Val < FDRcutoff]))` differentially expressed genes. Their chromosome distribution can be assessed.

```{r}
sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE)
plot(sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE))
axis(side=1, cex.axis=0.35)
```

This result also ressembles the analysis without surrogate variables, with a variable distribution in which big chromosomes rank higher, as well as some smaller chromosomes.

A list of the top 10 differentially expressed genes can be obtained:

```{r}
fit$genes$symbol[top]
```

It can be observed that the genes in the list have changed.

#### Diagnostics

The performance of the statistical test is diagnosed with the distributions of raw p-values and moderated t-statistics.

```{r echo=FALSE}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

The result is, again, similar to the previous analysis. The flat distribution for non-DE genes is present, but the peak is higher, as there are more differentially expressed genes. The QQ plot is similar.

The results are diagnosed with volcano and MA plots, in which the top 10 differentially expressed genes are highlighted.

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 3, 2))
volcanoplot(fit, coef = 2, highlight = 10, names = fit$genes$symbol, main = "Volcano")

top <- order(fit$lods[, 2], decreasing = TRUE)[1:10]
limma::plotMA(fit, coef = 2, status = rownames(fit$lods) %in% DEgenes, legend = FALSE,
main = "MA Plot", hl.pch = 46, hl.cex = 4, bg.pch = 46, bg.cex = 3, las = 1)
text(fit$Amean[top], fit$coef[top, 2], fit$genes$symbol[top], cex = 0.5, pos = 4)
```

### Performance comparison

```{r}

```

## Session information

```{r, message=FALSE}
sessionInfo()
```
