---
output:
  BiocStyle::html_document
---

<!---
Because we split the analysis pipeline in different independent files, to speed up processing it, here in the setup block we load libraries and objects that were loaded or produced in the previously processed file, and which are necessary in this file.
--->

```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(geneplotter)
opts_chunk$set(cache=TRUE,
               cache.path="cache/file2",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")
se.unpaired <- readRDS(file.path("rawCounts", "seLUSC.rds"))
dg.unpaired <- readRDS(file.path("results", "dge.rds"))
se <- readRDS(file.path("results", "se.paired.rds"))
dge <- readRDS(file.path("results", "dge.paired.rds"))
se.filt.unnorm <- readRDS(file.path("results", "se.paired.filt.unnorm.rds"))
dge.filt.unnorm <- readRDS(file.path("results", "dge.paired.filt.unnorm.rds"))
se.filt <- readRDS(file.path("results", "se.paired.filt.rds"))
dge.filt <- readRDS(file.path("results", "dge.paired.filt.rds"))
```

# Differential expression

## Differential expression using limma-voom

We perform a simple examination of expression changes and their associated p-values using the R/Bioconductor package [sva](http://bioconductor.org/packages/sva).

In order to use the method, a matrix of the model, including adjustment variables, and a null model which only includes the adjustment variables have to be created. In our case there are no factors to adjust, so we give ~1 as the intercept term in the model. Also, as the design of the analysis is paired (both types of samples come from the same patient), we need to add an identifier of the patient to the model.

Even if no known sources of variation were found, we can use SVA to estimate surrogate variables and add them to the model. These variables represent the unknown variation of the data.

```{r, message=FALSE}
library(sva)
# Obtain the patient ID, as the analysis is paired
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)

# Build the model (mod) and null (mod0) matrices 
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))

# Estimate surrogate variables
sv <- sva(assays(se.filt)$logCPM, mod = mod, mod0 = mod0)

#pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
#sum(p.adjust(pv, method="fdr") < 0.01)

# Add surrogates to the design matrix
len <- length(colnames(mod))
mod <- cbind(mod, sv$sv)
colnames(mod) <- c(colnames(mod)[1:len], paste0("SV", 1:sv$n))
v <- voom(dge.filt, mod, plot=TRUE)

fit <- lmFit(v, mod) # Fit the model with the voom weights
fit <- eBayes(fit) # Calculate moderated t-statistics
```

`sva` was able to estimate `r sv$n.sv` surrogate variables. Regarding the mean-variance plot obtained with `voom`, it shows the variance in the y axis and the mean of expression in the x axis for each gene, represented as a black dot. The red curve is fitted to the dots, and shows the trends in the data: increases in expression levels usually mean decreases in the variance, until a plateau is reached for highly expressed genes. We observe this behaviour in our plot. Moreover, the plot is also a diagnosis of filtering steps performed previously. The fact that there is not a drop in variance for the lower expression values indicates that the filtering threshold was good.

<!--Info about voom plot from https://f1000research.com/articles/5-1408, https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29 and https://stats.stackexchange.com/questions/160255/voom-mean-variance-trend-plot-how-to-interpret-the-plot-->

As we are performing multiple tests, the probability of type I errors (rejecting the null hypotheses when it is true). There are multiple statistical tools to decrease this errors, such as false discovery rate (FDR). This technique consists on defining an acceptable rate of type I errors (false discoveries). Here, we set a FDR of 0.05, which means that 5% of our rejections of the null hypothesis will be incorrect.

```{r}
FDRcutoff <- 0.05
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

Gene metadata such as the gene symbol and the chromosome can be added for an easier interpretation of the obtained statistics.

```{r}
# Get gene info...
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)

# and add it to the results
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 20)
```

What we can see from this table is that the FDR-adjusted p-values are close to 1, so they are not significant. We can check how many genes can be considered differentially expressed under the `r FDRcutoff` FDR cuttoff.

```{r}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
```

So our differential expression analysis results in `r length(rownames(tt)[tt$adj.P.Val < FDRcutoff]))` differentially expressed genes.

The results of a differential expression analysis can also be diagnosed with the distributions of raw p-values and moderated t-statistics.

```{r}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

Under the null hypothesis, we would expect a flat distribution of p-values. *And I think that the fact that there is not a enrichment of the low p-values is good, as this would mean that we are getting false positives because of too much variation in the data, or others.*

Regarding the second plot, it represents the quantiles of our data sample against the theoretical distribution they should follow. The diagonal represents the null hypothesis.

I'M LEAVING THE COMMENT OF THIS TWO PLOTS UNTIL SOMEBODY TELLS ME WHAT THEY ACTUALLY MEAN

## Enrichment analysis

```{r}
library(GSEABase)

geneUniverse <- rownames(se.filt)
length(geneUniverse)

c4 <- getGmt("c4.all.v6.2.entrez.gmt")
c4
length(c4)
```

```{r}
gsc <- GeneSetCollection(c(c4))
#map identifiers
gsc <- mapIdentifiers(gsc, AnnoOrEntrezIdentifier(metadata(se.filt)$annotation))
gsc
```

```{r}
Im <- incidence(gsc)
dim(Im)
Im[1:2, 1:10]
Im <- Im[, colnames(Im) %in% rownames(se.filt)]
dim(Im)
```

```{r}
se.filt <- se.filt[colnames(Im), ]
dim(se.filt)
dge.filt <- dge.filt[colnames(Im), ]
dim(dge.filt)
```

```{r}
library(limma)
library(sva)
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))


sv <- sva(assays(se.filt)$logCPM, mod, mod0)
#pv <- f.pvalue(assays(se.filt)$logCPM, mod, mod0)
#sum(p.adjust(pv, method="fdr") < 0.01)
# Add surrogates to the model
len <- length(colnames(mod))
mod <- cbind(mod, sv$sv)
colnames(mod) <- c(colnames(mod)[1:len], paste0("SV", 1:sv$n))
v <- voom(dge.filt, mod, plot=TRUE)
fit <- lmFit(v, mod)
fit <- eBayes(fit)
tt <- topTable(fit, coef = 2, n = Inf)
```

```{r}
Im <- Im[rowSums(Im) >= 5, ]
dim(Im)

tGSgenes <- tt[match(colnames(Im), rownames(tt)), "t"]
length(tGSgenes)

head(tGSgenes)

zS <- sqrt(rowSums(Im)) * (as.vector(Im %*% tGSgenes)/rowSums(Im))
length(zS)

head(zS)

rnkGS <- sort(abs(zS), decreasing = TRUE)
head(rnkGS)
```

```{r}
pv <- pmin(pnorm(zS), 1 - pnorm(zS))
pvadj <- p.adjust(pv, method = "fdr")
DEgs <- names(pvadj)[which(pvadj < 0.01)]
length(DEgs)
```

#GSEA DOESNT WORK

#GSVA
```{r}
library(GSVA)
GSexpr <- gsva(assays(se.filt)$logCPM, gsc, min.sz=5, max.sz=300, verbose=FALSE)
dim(GSexpr)
```
```{r}
mod <- model.matrix(~se.filt$type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~1, data = colData(se.filt))
svaobj <- sva(GSexpr, mod, mod0)

modSVs <- cbind(mod, svaobj$sv)

corfit <- duplicateCorrelation(GSexpr, modSVs, block = se.filt$cellline)
fit <- lmFit(GSexpr, modSVs)
fit <- eBayes(fit)
tt <- topTable(fit, coef = 2, n = Inf)
DEgs <- rownames(tt[tt$adj.P.Val < 0.01, , drop = FALSE])
length(DEgs)
```

```{r}
plot(tt$logFC, -log10(tt$P.Value), xlab="Log2 fold-change", ylab="-log10 P-value", 
     pch=".", cex=5, col=grey(0.75), cex.axis=1.2, cex.lab=1.5, las=1)
posx <- tt[tt$adj.P.Val < 0.01, "logFC"] 
posy <- -log10(tt[tt$adj.P.Val < 0.01, "P.Value"]) 
points(posx, posy, pch=".", cex=5, col="red")

```

```{r}
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 10)
```

```{r}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
saveRDS(DEgenes, file.path("results", "DEgenes.rds"))
```

```{r}
library(limma)
par(mfrow = c(1, 2), mar = c(4, 5, 3, 2))
volcanoplot(fit, coef = 2, highlight = 7, names = fit$genes$symbol, main = "Model")
top10 <- order(fit$lods[, 2], decreasing = TRUE)[1:10]
plotMA(fit, coef = 2, status = rownames(fit$lods) %in% DEgenes, legend = FALSE, 
    main = "Model 6", hl.pch = 46, hl.cex = 4, bg.pch = 46, bg.cex = 3, las = 1)
text(fit$Amean[top10], fit$coef[top10, 2], fit$genes$symbol[top10], cex = 0.5, pos = 4)
```












-----------------------------------------------------------
There are `r sum(p.adjust(pv, method="fdr") < 0.01)` genes changing significantly their expression at FDR < 1%. In Figure \@ref(fig:pdist) below we show the distribution of the resulting p-values. This distribution should be uniform with a peak at low p-values for the truly DE genes, assuming that most genes are not differentially expressed.

```{r pdist, echo=FALSE, out.width="800px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples."}
hist(pv, main="", las=1)

```

The `sva()` function is able to estimate surrogate variables, which have continuous values that can be used to correct and adjust unmeasured and unwanted effects.

```{r}
sv <- sva(assays(se.filt)$logCPM, mod, mod0)
sv$n
```

The SVA algorithm has found `r sv$n` surrogate variables. Let's use them to assess againt the extent of differential expression this time adjusting for these surrogate variables.

```{r}
modsv <- cbind(mod, sv$sv)
mod0sv <- cbind(mod0, sv$sv)
pvsv <- f.pvalue(assays(se.filt)$logCPM, modsv, mod0sv)
sum(p.adjust(pvsv, method="fdr") < 0.01)
```

We have increased the number of changing genes to `r sum(p.adjust(pvsv, method="fdr") < 0.01)`. Figure \@ref(fig:psvdist) shows the resulting distribution of p-values.

```{r psvdist, echo=FALSE, out.width="800px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples, adjusting for surrogate variables estimated with SVA."}
hist(pvsv, main="", las=1)
```

## Session information

```{r, message=FALSE}
sessionInfo()
```
