---
output:
  BiocStyle::html_document
---

<!---
Because we split the analysis pipeline in different independent files, to speed up processing it, here in the setup block we load libraries and objects that were loaded or produced in the previously processed file, and which are necessary in this file.
--->

```{r setup, cache=FALSE, echo=FALSE, results='hide', message=FALSE}
library(knitr)
library(SummarizedExperiment)
library(edgeR)
library(geneplotter)
opts_chunk$set(cache=TRUE,
               cache.path="cache/file2",
               cache.extra=R.version.string,
               autodep=TRUE,
               fig.align="center",
               comment="")
se.unpaired <- readRDS(file.path("rawCounts", "seLUSC.rds"))
dg.unpaired <- readRDS(file.path("results", "dge.rds"))
se <- readRDS(file.path("results", "se.paired.rds"))
dge <- readRDS(file.path("results", "dge.paired.rds"))
se.filt.unnorm <- readRDS(file.path("results", "se.paired.filt.unnorm.rds"))
dge.filt.unnorm <- readRDS(file.path("results", "dge.paired.filt.unnorm.rds"))
se.filt <- readRDS(file.path("results", "se.paired.filt.rds"))
dge.filt <- readRDS(file.path("results", "dge.paired.filt.rds"))
```

# Differential expression analysis

Differential expression analysis consists of identifying changes in gene expression. For RNAseq experiments, this means comparing relative concentrations of RNA molecules. In R, this can be done by using the [limma](http://bioconductor.org/packages/release/bioc/html/limma.html) and [sva](http://bioconductor.org/packages/sva.html) R/Bioconductor packages. 

## Analysis

`limma` allows for linear model analysis of data for DE. This requires building two model matrices: a matrix of the model including the outcome of interest and adjustment variables, and a matrix of the null model that only includes adjustment variables. The outcome of interest is the type of sample (tumor vs normal), and as both types of samples come from the same patient the analysis is paired, so a unique identifier for each patient is included. Finally, as there are unknown factors to adjust, an intercept term of 1 is given to the null model.

Even if no known sources of variation were found, surrogate variables representing unknown sources of variation can be added to the model. Those can be estimated with the `sva` package, and later be appended to the model. Finally, `voom` is used to fit the model as the library size distribution is not uniform across samples, and the moderated t-statistics are calculated. 

```{r mean-variance-SVA, message=FALSE, out.width="800px", fig.cap="Mean-variance plot for DE analysis with SVA."}
library(sva)
# Obtain the patient ID, as the analysis is paired
patientid <- substr(colnames(se.filt), 9, 12)
samplevial <- substr(colnames(se.filt), 14, 16)

# Build the model (mod) and null (mod0) matrices 
mod <- model.matrix(~type + patientid, data = colData(se.filt))
mod0 <- model.matrix(~patientid, data = colData(se.filt))

# Estimate surrogate variables
sv <- sva(assays(se.filt)$logCPM, mod = mod, mod0 = mod0)

# Add surrogates to the design matrix
len <- length(colnames(mod))
mod <- cbind(mod, sv$sv)
colnames(mod) <- c(colnames(mod)[1:len], paste0("SV", 1:sv$n))
v <- voom(dge.filt, mod, plot=TRUE)

fit <- lmFit(v, mod) # Fit the model with the voom weights
fit <- eBayes(fit) # Calculate moderated t-statistics
```

`sva` was able to estimate `r sv$n.sv` surrogate variables. Regarding the mean-variance plot shown in Figure \@ref(fig:mean-variance-SVA), which was obtained with `voom`, it shows the variance in the y axis and the mean of expression in the x axis for each gene, represented as a black dot. Such plots contain a red curve that is fitted to the dots and show trends in the data. Normally increases in expression levels mean a decrease in the variance, until a plateau is reached for highly expressed genes. This trend is observed in the plot obtained for this data. Moreover, the plot is also a diagnosis of filtering steps performed upstream. The fact that there is not a drop in variance for the lower expression values indicates that the filtering threshold was good.

<!--Info about voom plot from https://f1000research.com/articles/5-1408, https://genomebiology.biomedcentral.com/articles/10.1186/gb-2014-15-2-r29 and https://stats.stackexchange.com/questions/160255/voom-mean-variance-trend-plot-how-to-interpret-the-plot-->

As we are performing multiple tests, the probability of type I errors (rejecting the null hypotheses when it is true) increases. There are multiple statistical strategies to decrease the likelihood of these errors, such as the false discovery rate (FDR). This technique consists of defining an acceptable rate of type I errors (false discoveries). Here, a FDR of 0.01 is set, which means that 1% of the rejections of the null hypothesis will be incorrect. For our data set, which has 11872 observations, up to 119 genes can be false positives (FP).

```{r}
FDRcutoff <- 0.01
res <- decideTests(fit, p.value = FDRcutoff)
summary(res)
```

The summary table displays the upregulated, not significant, and downregulated genes for each variable in the model.

Gene metadata such as the gene symbol and the chromosome can be added for an easier interpretation of the obtained statistics, and to observe the chromosome distribution of the differentially expressed genes.

```{r}
# Get gene info...
genesmd <- data.frame(chr = as.character(seqnames(rowRanges(se.filt))), symbol = rowData(se.filt)[, 1], stringsAsFactors = FALSE)

# and add it to the results
fit$genes <- genesmd
tt <- topTable(fit, coef = 2, n = Inf)
head(tt, n = 20)
```

This table contains the relevant statistics used in the analysis, such as raw and adjusted p-values. In the case of the later ones, it can be seen some of their values are pretty low, which means that the result is not significant. The number of differentially expressed genes under the `r FDRcutoff` FDR cuttoff can be checked by applying a filter to this table.

```{r}
DEgenes <- rownames(tt)[tt$adj.P.Val < FDRcutoff]
length(DEgenes)
saveRDS(DEgenes, file.path("results", "DEgenes_FDRcut.rds"))
```

The differential expression analysis results in `r length(rownames(tt)[tt$adj.P.Val < FDRcutoff])`, with the following chromosome distribution.

```{r}
tt.sign <- tt[tt$adj.P.Val < FDRcutoff,]
DEgenes <- rownames(tt.sign[abs(tt.sign$logFC)>2,])
length(DEgenes)
saveRDS(DEgenes, file.path("results", "DEgenes.rds"))
```


```{r chromosomedistr, out.width="800px", fig.cap="Chromosome distribution of the DE genes for the analysis without SVA"}
tt$chr <- sub( "^chr([X|Y|0-9]+.*)", "\\1", tt$chr, perl = T)
tt$chr <- sub( "(.*)_.*_.*$", "\\1A", tt$chr, perl = T)
sort(table(tt$chr[tt$adj.P.Val < FDRcutoff]), decreasing = TRUE)
plot(sort(table(tt$chr[tt$adj.P.Val < FDRcutoff & !grepl(".*[A|v][1]*$", tt$chr, perl=TRUE)]), decreasing = TRUE), main = "DE chromosomes", ylab = "DE genes", xlab = "Chromosomes", cex.axis = 0.7)
``` 

The distribution is shown in Figure \@ref(fig:chromosomedistr), and it is pretty variable: some differentially expressed genes are located in big chromosomes, but smaller chromosomes are sometimes located higher in the ranking from what would be expected by its size.

A list of the top 10 most siginifant differentially expressed genes can be obtained:

```{r pdist, echo=FALSE, out.width="800px", fig.cap="Distribution of raw p-values for an F-test on every gene between tumor and normal samples."}
hist(pv, main="", las=1)
```{r}
top <- order(fit$lods[, 2], decreasing = TRUE)[1:10]
fit$genes$symbol[top]
```

## Diagnostics

The performance of the statistical test is diagnosed with the distributions of raw p-values and moderated t-statistics, as shown in Figure \@ref(fig:pvalues-noSV1).
The distribution of raw p-values and the QQ plot moderated t-statistics, displayed in Figure \@ref(fig:pvalues-tstats), can be used as a diagnosis of the tests performed.

```{r pvalues-tstats, out.width="800px", fig.cap="Raw p-values distribution and QQ plot of the moderated t-statistics for DE analysis."}
par(mfrow = c(1, 2), mar = c(4, 5, 2, 2))
hist(tt$P.Value, xlab = "Raw P-values", main = "", las = 1)
qqt(fit$t[, 2], df = fit$df.prior + fit$df.residual, main = "", pch = ".", cex = 3)
abline(0, 1, lwd = 2)
```

Under the null hypothesis, which is that most genes are not differentially expressed, the raw p-value distribution should be uniform with a peak at low p-values for the truly DE genes. This coincides with the observed plot.

Regarding the QQ plot, it represents the quantiles of our data sample against the theoretical distribution they should follow. The diagonal represents the null hypothesis. Most genes are off-diagonal, indicating differential expression. The fact tha most of the genes are far from the null hypothesis is a measure of the significance of the results.

The differential expression results are diagnosed with volcano and MA plots, in which the top 10 differentially expressed genes are highlighted as shown in Figure \@ref(fig:volcano-ma-plots). <!--TODO: expl plots-->

```{r volcano-ma-plots, out.width="800px", fig.cap="Volcano and MA plots for DE analysis without SVA."}
par(mfrow = c(1, 2), mar = c(4, 5, 3, 2))
volcanoplot(fit, coef = 2, highlight = 10, names = fit$genes$symbol, main = "Volcano Plot")

limma::plotMA(fit, coef = 2, status = rownames(fit$lods) %in% DEgenes, legend = FALSE,
main = "MA Plot", hl.pch = 46, hl.cex = 4, bg.pch = 46, bg.cex = 3, las = 1)
text(fit$Amean[top], fit$coef[top, 2], fit$genes$symbol[top], cex = 0.5, pos = 4)
```

## Session information

```{r, message=FALSE}
sessionInfo()
```
